File: tools/automation/src/log-buffer-evidence.ts
import { createHash } from "node:crypto";
import { access, chmod, mkdir, readFile, stat, writeFile } from "node:fs/promises";
import path from "node:path";

import {
  flashMonitorCommand,
  internalCommandSpec,
  type AutomationCategory,
  type LogBufferEvidence,
} from "./contracts.generated.js";
import {
  factoryImageDigest,
  flashChildFailureFacts,
  flashEffectEnvironment,
  inspectFlashEffect,
} from "./flash-child-diagnostics.js";
import {
  fetchJsonFromSameOrigin,
  fetchTextResponseFromSameOrigin,
  type SameOriginTextResponse,
  uniqueRuntimeOrigin,
} from "./http.js";
import type { ProcessOutcome, ProcessPort } from "./process.js";
import { verifySemanticEvidenceRedaction } from "./redaction.js";
import { hasPassiveSafeState } from "./version-evidence.js";
import {
  captureTextWebSocketFrame,
  WebSocketProtocolError,
  type WebSocketFactory,
} from "./websocket.js";
import { assertWithinWorkspace } from "./workspace.js";

export type LogBufferEvidenceOptions = {
  readonly privateRoot: string;
  readonly packageManifest: string;
  readonly wifiCredentials: string;
  readonly port: string;
  readonly projection: string;
  readonly captureTimeoutSeconds: number;
};

type JsonObject = Readonly<Record<string, unknown>>;
type FailureCategory = Extract<
  AutomationCategory,
  "hardware_blocked" | "evidence_invalid" | "timeout" | "process_failed"
>;

const rawLogMarker = "axeos_websocket_logs=connected\n";
const expectedContentType = "text/plain";
const expectedContentDisposition = "attachment; filename=\"bitaxe-logs.txt\"";

export class LogBufferEvidenceError extends Error {
  public constructor(
    public readonly category: FailureCategory,
    message: string,
    public readonly publicValue: Readonly<Record<string, unknown>>,
  ) {
    super(message);
    this.name = "LogBufferEvidenceError";
  }

  public withCompletedFlash(): LogBufferEvidenceError {
    return new LogBufferEvidenceError(this.category, this.message, {
      ...this.publicValue,
      flash_effect_completed: true,
    });
  }
}

function failure(
  category: FailureCategory,
  message: string,
  facts: Readonly<Record<string, unknown>> = {},
): LogBufferEvidenceError {
  return new LogBufferEvidenceError(category, message, {
    stage: "log_buffer_capture",
    flash_effect_completed: false,
    ...facts,
  });
}

function sha256(value: string | Buffer): string {
  return createHash("sha256").update(value).digest("hex");
}

function object(value: unknown, context: string): JsonObject {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw failure("evidence_invalid", `${context} must be an object`);
  }
  return value as JsonObject;
}

function requiredString(value: JsonObject, field: string, context: string): string {
  const maybeValue = value[field];
  if (typeof maybeValue !== "string" || maybeValue === "") {
    throw failure("evidence_invalid", `${context} field is invalid`);
  }
  return maybeValue;
}

function monitorBootSession(document: string): string {
  const sessions = new Set<string>();
  for (const match of document.matchAll(/\bruntime_boot_identity session=([0-9a-f]{32})\b/gu)) {
    const maybeSession = match[1];
    if (maybeSession !== undefined) sessions.add(maybeSession);
  }
  if (sessions.size !== 1) {
    throw failure("hardware_blocked", "monitor capture lacks one stable boot session");
  }
  const [session] = sessions;
  if (session === undefined) {
    throw failure("hardware_blocked", "monitor capture boot session is missing");
  }
  return session;
}

function validateRuntimeIdentity(
  systemInfo: JsonObject,
  manifest: JsonObject,
  expectedBootSession: string,
): void {
  for (const [wireField, manifestField] of [
    ["sourceCommit", "source_commit"],
    ["referenceCommit", "reference_commit"],
    ["appElfSha256", "app_elf_sha256"],
  ] as const) {
    if (
      requiredString(systemInfo, wireField, "system info")
      !== requiredString(manifest, manifestField, "package manifest")
    ) {
      throw failure("evidence_invalid", "runtime identity does not match the exact package");
    }
  }
  if (requiredString(systemInfo, "bootSession", "system info") !== expectedBootSession) {
    throw failure("evidence_invalid", "runtime boot session does not match the admitted monitor session");
  }
}

function markerCount(document: string): number {
  return document.split(/\r?\n/gu).filter((line) => line === rawLogMarker.trimEnd()).length;
}

function headersMatch(response: SameOriginTextResponse): boolean {
  return response.contentType === expectedContentType
    && response.contentDisposition === expectedContentDisposition;
}

function validateCorrelation(
  baseline: SameOriginTextResponse,
  final: SameOriginTextResponse,
  frame: string,
): { readonly baselineMarkerCount: number; readonly finalMarkerCount: number } {
  if (!headersMatch(baseline) || !headersMatch(final)) {
    throw failure("evidence_invalid", "retained log download headers do not match");
  }
  if (frame !== rawLogMarker) {
    throw failure("evidence_invalid", "raw log WebSocket frame does not match the exact marker");
  }
  if (!final.body.startsWith(baseline.body)) {
    throw failure("evidence_invalid", "final retained log does not preserve the baseline prefix");
  }
  const baselineMarkerCount = markerCount(baseline.body);
  const finalMarkerCount = markerCount(final.body);
  if (finalMarkerCount !== baselineMarkerCount + 1) {
    throw failure("evidence_invalid", "retained log does not contain exactly one new marker");
  }
  return { baselineMarkerCount, finalMarkerCount };
}

async function createPrivateRoot(root: string): Promise<void> {
  try {
    await stat(root);
    throw failure("evidence_invalid", "private attempt root must be absent before launch");
  } catch (error) {
    if (error instanceof LogBufferEvidenceError) throw error;
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }
  await mkdir(root, { recursive: true, mode: 0o700 });
  await chmod(root, 0o700);
}

async function privateJson(output: string, value: unknown): Promise<void> {
  await writeFile(output, `${JSON.stringify(value, null, 2)}\n`, {
    encoding: "utf8",
    flag: "wx",
    mode: 0o600,
  });
  await chmod(output, 0o600);
}

async function child(
  processPort: ProcessPort,
  program: string,
  args: readonly string[],
  context: string,
): Promise<ProcessOutcome> {
  try {
    return await processPort.run(internalCommandSpec(program, [...args], (value) => value));
  } catch {
    throw failure("process_failed", `${context} launch failed`);
  }
}

async function systemInfo(origin: URL, output: string): Promise<JsonObject> {
  try {
    return object(await fetchJsonFromSameOrigin(origin, "/api/system/info", output), "system info");
  } catch (error) {
    if (error instanceof LogBufferEvidenceError) throw error;
    throw failure("hardware_blocked", "same-origin system info request failed");
  }
}

async function retainedLog(origin: URL, output: string): Promise<SameOriginTextResponse> {
  try {
    return await fetchTextResponseFromSameOrigin(origin, "/api/system/logs", output);
  } catch {
    throw failure("hardware_blocked", "same-origin retained log request failed");
  }
}

async function rawLogFrame(
  origin: URL,
  output: string,
  maybeWebSocketFactory: WebSocketFactory | undefined,
): Promise<string> {
  try {
    return await captureTextWebSocketFrame(origin, "/api/ws", output, maybeWebSocketFactory);
  } catch (error) {
    if (error instanceof WebSocketProtocolError) {
      throw failure("evidence_invalid", "raw log WebSocket frame is invalid");
    }
    throw failure("hardware_blocked", "same-origin raw log WebSocket request failed");
  }
}

export async function captureLogBufferEvidence(
  workspaceRoot: string,
  options: LogBufferEvidenceOptions,
  processPort: ProcessPort,
  flashProgram: string,
  validatorProgram: string,
  maybeWebSocketFactory?: WebSocketFactory,
): Promise<LogBufferEvidence> {
  const privateRoot = assertWithinWorkspace(workspaceRoot, options.privateRoot);
  const manifestPath = assertWithinWorkspace(workspaceRoot, options.packageManifest);
  const credentialsPath = assertWithinWorkspace(workspaceRoot, options.wifiCredentials);
  const projectionPath = assertWithinWorkspace(workspaceRoot, options.projection);
  await access(manifestPath);
  await access(credentialsPath);
  await createPrivateRoot(privateRoot);

  const manifestDocument = await readFile(manifestPath, "utf8");
  let manifest: JsonObject;
  let factoryDigest: string;
  try {
    manifest = object(JSON.parse(manifestDocument), "package manifest");
    requiredString(manifest, "source_commit", "package manifest");
    requiredString(manifest, "reference_commit", "package manifest");
    requiredString(manifest, "app_elf_sha256", "package manifest");
    factoryDigest = factoryImageDigest(manifest);
  } catch (error) {
    if (error instanceof LogBufferEvidenceError) throw error;
    throw failure("evidence_invalid", "package manifest identity is invalid");
  }

  const manifestDigest = sha256(manifestDocument);
  const effectPath = path.join(privateRoot, "flash-effect.private.json");
  const expectedEffectIdentity = {
    packageIdentityDigest: manifestDigest,
    factoryImageDigest: factoryDigest,
  };
  const baseSpec = flashMonitorCommand(flashProgram, {
    board: 205,
    port: options.port,
    manifest: manifestPath,
    wifiCredentials: credentialsPath,
    captureTimeoutSeconds: options.captureTimeoutSeconds,
    evidenceMode: "dual",
    evidenceDir: privateRoot,
  });
  const flashSpec = internalCommandSpec(
    baseSpec.program,
    [...baseSpec.args],
    baseSpec.result,
    flashEffectEnvironment(effectPath, expectedEffectIdentity),
  );
  let flashOutcome: ProcessOutcome;
  try {
    flashOutcome = await processPort.run(flashSpec);
  } catch {
    const effect = await inspectFlashEffect(effectPath, expectedEffectIdentity);
    throw failure(
      "process_failed",
      "exact-package flash-monitor launch failed",
      flashChildFailureFacts(undefined, effect),
    );
  }
  const effect = await inspectFlashEffect(effectPath, expectedEffectIdentity);
  const effectFacts = flashChildFailureFacts(flashOutcome, effect);
  if (flashOutcome.timedOut) {
    throw failure("timeout", "exact-package flash-monitor timed out", effectFacts);
  }
  if (flashOutcome.exitCode !== 0) {
    throw failure("hardware_blocked", "exact-package flash-monitor did not reach readiness", effectFacts);
  }
  if (effect.flash_effect_result_status !== "valid" || effect.flash_effect_status !== "completed") {
    throw failure("evidence_invalid", "exact-package flash effect result is invalid", effectFacts);
  }

  try {
    const monitor = await readFile(path.join(privateRoot, "flash-monitor.classifier-input.log"), "utf8");
    if (!hasPassiveSafeState(monitor)) {
      throw failure("hardware_blocked", "boot lacks passive safe-state evidence");
    }
    const bootSession = monitorBootSession(monitor);
    let origin: URL;
    try {
      origin = uniqueRuntimeOrigin(monitor);
    } catch {
      throw failure("hardware_blocked", "runtime origin admission is invalid");
    }
    const runtime = await systemInfo(origin, path.join(privateRoot, "system-info.private.json"));
    validateRuntimeIdentity(runtime, manifest, bootSession);
    const baseline = await retainedLog(origin, path.join(privateRoot, "baseline-log.private.txt"));
    const frame = await rawLogFrame(
      origin,
      path.join(privateRoot, "raw-log-frame.private.txt"),
      maybeWebSocketFactory,
    );
    const final = await retainedLog(origin, path.join(privateRoot, "final-log.private.txt"));
    const counts = validateCorrelation(baseline, final, frame);
    const evidence: LogBufferEvidence = {
      schema_version: "bitaxe-log-buffer-evidence-v1",
      board: 205,
      source_commit: requiredString(manifest, "source_commit", "package manifest"),
      reference_commit: requiredString(manifest, "reference_commit", "package manifest"),
      package_manifest_sha256: manifestDigest,
      workflow: {
        schema_version: "bitaxe-workflow-identity-v1",
        command: "capture-log-buffer-evidence",
        request_sha256: sha256(JSON.stringify({
          manifest: manifestDigest,
          routes: ["/api/system/info", "/api/system/logs", "/api/ws"],
          timeout: options.captureTimeoutSeconds,
        })),
      },
      detector_admitted: true,
      boot_observed: true,
      same_origin_observed: true,
      log_buffer: {
        boot_session_sha256: sha256(bootSession),
        baseline_body_sha256: sha256(baseline.body),
        final_body_sha256: sha256(final.body),
        raw_frame_sha256: sha256(frame),
        baseline_bytes: Buffer.byteLength(baseline.body, "utf8"),
        final_bytes: Buffer.byteLength(final.body, "utf8"),
        raw_frame_bytes: Buffer.byteLength(frame, "utf8"),
        baseline_marker_count: counts.baselineMarkerCount,
        final_marker_count: counts.finalMarkerCount,
        new_marker_count: 1,
        both_download_headers_match: true,
        baseline_is_exact_prefix: true,
        raw_frame_plain_text: true,
        raw_frame_marker_matches: true,
        retained_marker_matches_frame: true,
      },
      mining_state: "disabled",
      hardware_control_state: "disabled",
      cleanup_complete: true,
      redaction_status: "passed",
    };
    const privateCandidate = path.join(privateRoot, "final-evidence.private.json");
    await privateJson(privateCandidate, evidence);
    await verifySemanticEvidenceRedaction(privateRoot);
    const validation = await child(
      processPort,
      validatorProgram,
      [privateCandidate],
      "log buffer evidence validator",
    );
    if (validation.timedOut) {
      throw failure("timeout", "log buffer evidence validation timed out");
    }
    if (validation.exitCode !== 0) {
      throw failure("evidence_invalid", "log buffer evidence validation failed");
    }
    await mkdir(path.dirname(projectionPath), { recursive: true });
    await writeFile(projectionPath, `${JSON.stringify(evidence, null, 2)}\n`, {
      encoding: "utf8",
      flag: "wx",
    });
    return evidence;
  } catch (error) {
    const primary = error instanceof LogBufferEvidenceError
      ? error
      : failure("evidence_invalid", "log buffer orchestration evidence is invalid");
    throw primary.withCompletedFlash();
  }
}

File: crates/bitaxe-automation-contracts/src/log_buffer_evidence.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::{AutomationCommand, WorkflowIdentity, LOG_BUFFER_EVIDENCE_SCHEMA};

const MAX_RETAINED_BODY_BYTES: u64 = 1024 * 1024;
const MAX_RAW_FRAME_BYTES: u64 = 256 * 1024;

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct LogBufferObservationEvidence {
    pub boot_session_sha256: String,
    pub baseline_body_sha256: String,
    pub final_body_sha256: String,
    pub raw_frame_sha256: String,
    pub baseline_bytes: u64,
    pub final_bytes: u64,
    pub raw_frame_bytes: u64,
    pub baseline_marker_count: u64,
    pub final_marker_count: u64,
    pub new_marker_count: u64,
    pub both_download_headers_match: bool,
    pub baseline_is_exact_prefix: bool,
    pub raw_frame_plain_text: bool,
    pub raw_frame_marker_matches: bool,
    pub retained_marker_matches_frame: bool,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct LogBufferEvidence {
    pub schema_version: String,
    pub board: u16,
    pub source_commit: String,
    pub reference_commit: String,
    pub package_manifest_sha256: String,
    pub workflow: WorkflowIdentity,
    pub detector_admitted: bool,
    pub boot_observed: bool,
    pub same_origin_observed: bool,
    pub log_buffer: LogBufferObservationEvidence,
    pub mining_state: String,
    pub hardware_control_state: String,
    pub cleanup_complete: bool,
    pub redaction_status: String,
}

impl LogBufferEvidence {
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.schema_version != LOG_BUFFER_EVIDENCE_SCHEMA || self.board != 205 {
            return Err("log buffer evidence schema or board is invalid");
        }
        if self.workflow.schema_version != "bitaxe-workflow-identity-v1"
            || self.workflow.command != AutomationCommand::CaptureLogBufferEvidence
        {
            return Err("log buffer workflow identity is invalid");
        }
        for commit in [self.source_commit.as_str(), self.reference_commit.as_str()] {
            if !is_lower_hex(commit, 40) {
                return Err("log buffer source identity is invalid");
            }
        }
        for digest in [
            self.package_manifest_sha256.as_str(),
            self.workflow.request_sha256.as_str(),
            self.log_buffer.boot_session_sha256.as_str(),
            self.log_buffer.baseline_body_sha256.as_str(),
            self.log_buffer.final_body_sha256.as_str(),
            self.log_buffer.raw_frame_sha256.as_str(),
        ] {
            if !is_lower_hex(digest, 64) {
                return Err("log buffer evidence digest is invalid");
            }
        }
        let observation = &self.log_buffer;
        let Some(minimum_final_bytes) = observation
            .baseline_bytes
            .checked_add(observation.raw_frame_bytes)
        else {
            return Err("log buffer observation is incomplete");
        };
        let Some(expected_final_marker_count) = observation.baseline_marker_count.checked_add(1)
        else {
            return Err("log buffer observation is incomplete");
        };
        if observation.baseline_bytes > MAX_RETAINED_BODY_BYTES
            || observation.final_bytes > MAX_RETAINED_BODY_BYTES
            || observation.raw_frame_bytes == 0
            || observation.raw_frame_bytes > MAX_RAW_FRAME_BYTES
            || observation.final_bytes < minimum_final_bytes
            || observation.final_marker_count != expected_final_marker_count
            || observation.new_marker_count != 1
            || !observation.both_download_headers_match
            || !observation.baseline_is_exact_prefix
            || !observation.raw_frame_plain_text
            || !observation.raw_frame_marker_matches
            || !observation.retained_marker_matches_frame
        {
            return Err("log buffer observation is incomplete");
        }
        if !self.detector_admitted
            || !self.boot_observed
            || !self.same_origin_observed
            || self.mining_state != "disabled"
            || self.hardware_control_state != "disabled"
            || !self.cleanup_complete
            || self.redaction_status != "passed"
        {
            return Err("log buffer safety or privacy evidence is invalid");
        }
        Ok(())
    }
}

fn is_lower_hex(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

#[cfg(test)]
mod tests {
    use super::{LogBufferEvidence, LogBufferObservationEvidence};
    use crate::{AutomationCommand, WorkflowIdentity};

    fn valid_evidence() -> LogBufferEvidence {
        LogBufferEvidence {
            schema_version: "bitaxe-log-buffer-evidence-v1".to_owned(),
            board: 205,
            source_commit: "a".repeat(40),
            reference_commit: "b".repeat(40),
            package_manifest_sha256: "c".repeat(64),
            workflow: WorkflowIdentity {
                schema_version: "bitaxe-workflow-identity-v1".to_owned(),
                command: AutomationCommand::CaptureLogBufferEvidence,
                request_sha256: "d".repeat(64),
            },
            detector_admitted: true,
            boot_observed: true,
            same_origin_observed: true,
            log_buffer: LogBufferObservationEvidence {
                boot_session_sha256: "e".repeat(64),
                baseline_body_sha256: "f".repeat(64),
                final_body_sha256: "1".repeat(64),
                raw_frame_sha256: "2".repeat(64),
                baseline_bytes: 100,
                final_bytes: 150,
                raw_frame_bytes: 32,
                baseline_marker_count: 2,
                final_marker_count: 3,
                new_marker_count: 1,
                both_download_headers_match: true,
                baseline_is_exact_prefix: true,
                raw_frame_plain_text: true,
                raw_frame_marker_matches: true,
                retained_marker_matches_frame: true,
            },
            mining_state: "disabled".to_owned(),
            hardware_control_state: "disabled".to_owned(),
            cleanup_complete: true,
            redaction_status: "passed".to_owned(),
        }
    }

    #[test]
    fn valid_closed_projection_is_accepted() {
        // Arrange
        let evidence = valid_evidence();

        // Act
        let result = evidence.validate();

        // Assert
        assert_eq!(result, Ok(()));
    }

    #[test]
    fn missing_new_retained_marker_is_rejected() {
        // Arrange
        let mut evidence = valid_evidence();
        evidence.log_buffer.final_marker_count = 2;

        // Act
        let result = evidence.validate();

        // Assert
        assert_eq!(result, Err("log buffer observation is incomplete"));
    }
}
